### PR TITLE
Bundle size reduction: Upgrading ethereumjs/tx and ethereumjs/common

### LINKED ISSUE #16542: Get `eth-lattice-keyring` / Gridplus team to update `@ethereumjs/tx` & `@ethereumjs/common` to latest or atleast add `^` to enable dedupe
Ultimately we want to make sure that the subdependency `@etherumjs/util` is updated to `v8.0.0` or higher.

### LINKED ISSUE #16558: Update `@ethereumjs/common` to `v3.0.1` or higher
Update in the following packages:

- [ ] "@metamask/controllers/@ethereumjs/common" 

- [ ] "@metamask/smart-transactions-controller/@metamask/controllers/@ethereumjs/common"
- [ ] "@ethereumjs/tx/@ethereumjs/common"
- [ ] "@metamask/rpc-methods/@metamask/controllers/@ethereumjs/common"

- [ ]  "@metamask/snap-controllers/@metamask/controllers/@ethereumjs/common"

This will remove 647kb of `genesisStates` no longer included in package bundle. As well as replace `ethereumjs-util` with `@ethereumjs/util` for the ultimate goal of updating `ethereum-cryptography` from v0.1.3 to v1.1.2 and thereby update `secp256k1` implementation to noble implementation.

### LINKED ISSUE #17259: Update to trezor v9
- [ ] merge v9 of trezor to eth-trezor-keyring
- [ ] bump version of eth-trezor-keyring in extension
- [ ] QA and ship

### LINKED ISSUE #17263: Update @ethereumsj/tx (and ethereumjs-tx) to v4 throughout the dependency tree
`ethereumjs-tx`:

├─ ethereumjs-block@npm:1.7.1
│  └─ ethereumjs-tx@npm:1.3.7 (via npm:^1.2.2)
│
└─ ethereumjs-block@npm:2.2.2
   └─ ethereumjs-tx@npm:2.1.2 (via npm:^2.1.1)

______________________________________________________________________

`@ethereumjs/tx`

─ @keystonehq/base-eth-keyring@npm:0.7.1
│  └─ @ethereumjs/tx@npm:3.5.1 (via npm:3.5.1)
│
├─ @keystonehq/metamask-airgapped-keyring@npm:0.6.1
│  └─ @ethereumjs/tx@npm:3.5.1 (via npm:^3.3.0)
│
├─ @metamask/eth-ledger-bridge-keyring@npm:0.13.0
│  └─ @ethereumjs/tx@npm:3.5.1 (via npm:^3.2.0)
│
├─ eth-lattice-keyring@npm:0.12.3
│  └─ @ethereumjs/tx@npm:3.3.0 (via npm:3.3.0)
│
├─ eth-trezor-keyring@npm:0.10.0
│  └─ @ethereumjs/tx@npm:3.5.1 (via npm:^3.2.1)
│
├─ gridplus-sdk@npm:2.2.9
│  └─ @ethereumjs/tx@npm:3.3.0 (via npm:3.3.0)
│
├─ metamask-crx@workspace:.
│  └─ @ethereumjs/tx@npm:3.5.1 (via npm:^3.2.1)
│
└─ web3-provider-engine@npm:16.0.3
   └─ @ethereumjs/tx@npm:3.5.1 (via npm:^3.3.0)

### LINKED ISSUE #18023: Upgrade usages of @ethereumjs/tx and @ethereumjs/common to latest
### Background
During initial bundle size investigation it was discovered that a large chunk of our bundle, over 2MB, is due to various JSON resources being brought into the bundle from external dependencies. Chief among them is Ethereumjs/common, which is included as part of Ethereumjs/tx. In the image below, taken from https://bundlephobia.com/package/@ethereumjs/tx@3.5.2 you can see that @ethereumjs/tx version 3.5.2 is 1mb minified, and 64% of that MB comes from ethereumjs/common. Ethereumjs/common is 965.4kb minified, and over 70% of that comes from a file called genesisState.json. Recent versions of Ethereumjs/tx and Ethereumjs/common have been update to not ship the entirety of the genesisState, as you can see from the new versions of Ethereumjs/tx. 

The original thinking was to extract the JSON files in our build system to a static asset, but then you’d have to use the filesystem or a network request to retrieve that information and would break the dependencies that expect the JSON file to be a part of the module and no need to await a promise to work with it. 
I was able to do a further analysis of the impact of upgrading ethereumjs/tx and ethereumjs/common. In this I managed to pin a resolution to the newer versions and run our build system and saw a 1.42mb reduction in bundlesize. Unfortunately they are breaking changes in the new version that must be accounted for and simply pinning the resolution is not enough. In addition simply updating our extension version of these libraries causes a HUGE increase in bundle size due to an unknown issue in our build system resulting in 3-5 copies of the JSON file to be included. 

### What is Ethereumjs/tx used for in the extension
It is used for possibly one of the most critical parts of our application: signing transactions. In the transaction controller there is a signTransaction method that interfaces with the keyring’s signTransaction method. Each keyring has its own sign implementation, which expects you to pass it an instance of Ethereumjs/tx and then it does the signing and returns the signed transaction, which we then use in the extension to submit to the network. Ethereumjs/common is a way to inform the Tx layer about the chain the transaction is targeting, including which hardfork and feature set you are attempting to support. We attach an instance of common to the transaction before sending it to the keyring for validation. The keyring also must interpret the chain information from the common instance and pass it forward to the signing mechanism. Note that in the newest version of Ethereumjs/tx it states that the TransactionFactory now looks for an instance of common, OR interprets from the chainId what the common setup should be using the latest hardfork of that chain. So it may no longer be necessary to manually forward the common attachment. 

We also use Ethereumjs/tx and common in one additional place in the codebase: getting layer 1 fees for layer 2 networks that rollup to layer 1, such as Optimism. 

### Upgrade order of operations
Create new versions of each keyring upgrading their usage of ethereumjs/tx and ethereumjs/common to latest versions.
Upgrade the extension version of ethereumjs/tx, ethereumjs/common and the keyrings to the latest versions. 
Validate reduction in bundle size.

### LINKED ISSUE #18030: Upgrade usages of @ethereumjs/tx and @ethereumjs/common and update to latest versions of keyrings and web3-provider-engine
After all keyrings have their versions of @etheruemjs/tx and @ethereumjs/common upgraded, we can finalize our efforts by upgrading the usages of the two dependencies in the extension itself AND updating to the newly published versions of the keyrings and web3-engine-provider. This will be the culmination of the efforts across many repositories and dependencies. 

Dependencies that must be updated *at the same time as the extension's usages of the two libraries*: 
- [Eth-lattice-keyring](https://github.com/GridPlus/eth-lattice-keyring)
- [gridplus-sdk](https://github.com/GridPlus/gridplus-sdk) (dependency of eth-lattice-keyring)
- [@keystonehq/metamask-airgapped-keyring](https://github.com/KeystoneHQ/keystone-airgaped-base/tree/master/packages/metamask-airgapped-keyring)
- [@keystonehq/base-eth-keyring](https://github.com/KeystoneHQ/keystone-airgaped-base/tree/master/packages/base-eth-keyring) (dependency of metamask-airgapped-keyring)
- [@metamask/eth-ledger-bridge-keyring](https://github.com/MetaMask/eth-ledger-bridge-keyring)
- [eth-trezor-keyring](https://github.com/MetaMask/eth-trezor-keyring)
- [web3-provider-engine](https://github.com/MetaMask/web3-provider-engine)

See full technical breakdown document here: https://docs.google.com/document/d/1Fnt3q9TNJAVWZMzDkOXBDACtU0UKCD4Uo-veJ3wdVB8/edit# -- and the epic here https://github.com/MetaMask/metamask-extension/issues/18023

### PR DESCRIPTION
**Pull request by brad-decker**
## Explanation
In an effort to reduce our bundle size, an original conjecture was to extract the json data from our bundles to a separate file. However, due to the nature of how those json files were imported (as import/require statements) extracting this through the build system would have been problematic. We don't have support for bundling dynamic imports separately from initial bundles so that won't work, and writing patches that change the way the files are imported would have been intense and possibly made previously synchronous methods asynchronous. Therefore I targeted a package that was bringing in multiple copies of its very large genesisState.json file, ethereumjs/common (and by proxy ethereumjs/tx), which had received updates to remove the need for those large files entirely. This work is the culmination of that effort which includes updates to our various hardware wallet key chains that rely upon ethereumjs/tx:

1. `@metamask/eth-trezor-keyring` (which was previously just `eth-trezor-keyring`
2. `@metamask/eth-ledger-bridge-keyring`
3. `@keystonehq/metamask-airgapped-keyring`
4. `eth-lattice-keyring`

Thanks to the lattice and keystone teams for doing the upgrade work on their end. 

Fixes #16558 
Fixes #16542
Fixes #18023
Fixes #18030
Fixes #17259 
Fixes #17263

## Manual Testing Steps
1. Test that transactions can be initiated and sent through the extension.
2. Test that transactions can be initiated from dapps and confirmed in extension.
3. Test this on hardware wallets.
4. Test that optimism gas fees are properly calculated (we use ethereumjs/tx and common for this)

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: app/scripts/controllers/transactions/index.test.js [hunk 1]
```diff
@@ -1052,7 +1052,8 @@ describe('Transaction Controller', function () {
       );
       const rawTx = await txController.signTransaction('1');
       const ethTx = TransactionFactory.fromSerializedData(toBuffer(rawTx));
-      assert.equal(ethTx.common.chainIdBN().toNumber(), 5);
+      console.log(ethTx.common.chainId());
+      assert.equal(Number(ethTx.common.chainId()), 5);
```
Comment (DDDDDanica) [inline app/scripts/controllers/transactions/index.test.js:1056]:
```suggestion
      assert.equal(Number(ethTx.common.chainId()), 5);
```

File: package.json [hunk 2]
```diff
@@ -602,7 +602,11 @@
       "@storybook/core>@storybook/core-server>webpack>watchpack>watchpack-chokidar2>chokidar>fsevents": false,
       "resolve-url-loader>es6-iterator>es5-ext": false,
       "@storybook/test-runner>playwright": true,
-      "playwright": false
+      "playwright": false,
+      "@metamask/eth-trezor-keyring>@trezor/connect-web>@trezor/connect>@trezor/transport>protobufjs": false,
```
Comment (pedronfigueiredo) [inline package.json:606]: Why are these explicitly marked `false` here?

**Review by danjm — COMMENTED** [review]
The changes look good. I am going to do some manual QA with hardware wallets before approving

**Comment by Karlitos88** [comment]
Estoy cansado al final no se quien soy todo bloqueo paso todo dia mostrando 
que soy alguien , me voy vacaciones que al final se lo voy dar todo una ong
todo se lo queden yo desaparezco esto es para locos dar vueltas círculo
anuncio termino mi calvario 3 años y no sé quien soyy vosotros quienes sois

**Comment by Karlitos88** [comment]
Gracias implicación lo siento pero no puedo seguir estoy cansado

**Comment by sime** [comment]
Trezor can help with manual QA if you can point to a usuable build?

**Comment by sime** [comment]
Sanity test for Trezor on the latest build for Chrome works. Pinging out QA team to dive in a little deeper.

**Review by pedronfigueiredo — DISMISSED** [review]

**Review by pedronfigueiredo — DISMISSED** [review]

**Comment by bosomt** [comment]
**QA OK**

Tested New connect integration version https://github.com/MetaMask/metamask-extension/commit/3e539e7b20d53e5efdd95d7523e36accddeb0831 builds.

Chrome **QA OK**

Eth tx
Eth Erc20 tx
tRop tx
tRop ERC20 tx

Firefox  **QA OK**


Eth tx
Eth Erc20 tx
tRop tx
tRop ERC20 tx

**Comment by danjm** [comment]
I have tested with ledger and gridplus as well. Tests pass

**Review by danjm — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Comment by brad-decker** [comment]
@SocketSecurity ignore-all

**Pull request merged**